salt/_modules/caasp_kubernetes_crds.py

```diff
@@ -0,0 +1,62 @@
+from __future__ import absolute_import
+
+from salt.serializers.yaml import serialize, deserialize
+
+
+def __virtual__():
+    return "caasp_kubernetes_crds"
+
+
+def kubectl(command, **kwargs):
+    return __salt__['caasp_kubernetes_kubectl.kubectl'](command, **kwargs)
+
+
+def custom_resource_present(custom_resource, namespace='kube-system'):
+    ret, _, _ = kubectl('apply -f -', stdin=serialize(custom_resource, default_flow_style=False), namespace=namespace)
+    return ret == 0
+
+
+def custom_resource_missing(custom_resource, namespace='kube-system'):
+    ret, _, _ = kubectl('delete -f -', stdin=serialize(custom_resource, default_flow_style=False), namespace=namespace)
+    return ret == 0
+
+
+def fetch_current_custom_resources(custom_resource_name, namespace='kube-system'):
+    _, current_resources, _ = kubectl('get %s' % custom_resource_name, namespace=namespace, output='yaml')
+    return deserialize(current_resources)
+
+
+def reconcile_desired_resources(custom_resource_name, namespace, desired_resources=[]):
```

Comment:
I think this function tries to be too smart. I would mimic many other Salt modules and have a `caasp_kubernetes_crd.absent` (equivalent to [`file.missing`](https://docs.saltstack.com/en/latest/ref/states/all/salt.states.file.html#salt.states.file.absent)) and `caasp_kubernetes_crd.managed` (equivalent to [`file.managed`](https://docs.saltstack.com/en/latest/ref/states/all/salt.states.file.html#salt.states.file.absent)). Then we could skip all this code for detecting the current CRD and if we want to delete it or not...

Reply:
I agree it tries to be too smart, however I don't see a way around it. When a user modifies the settings on the UI and we run an orchestration we know what settings we want, and for removing the extra settings we need to reconcile somehow with the real state of the cluster and that needs to happen somewhere. Am I missing an alternative approach?

Reply:
I think we have two cases:

1) the _cleanup_ of registries not present in the pillar: for each registry obtained with `kubectl get registries`, if it is not managed by us, do a `caasp_kubernetes.crd.absent` (a `kubectl delete` under the hood). This could be done with a Jinja loop.
2) the `add/modify` registry. It would be handled by `kubectl apply`s (I think it doesn't really matter if we are creating the CR for the first time or if we are modifying the resource: the k8s API will deal with updating it if that is the case). We could iterate for all the certificates in `get_certs()` and  `caasp_kubernetes.crd.manage` each one. This could be done with a Jinja loop.

Reply:
Okay, so moving a bit of the logic to the rendering. I'm not sure I'm very happy with that, because with the current solution the SLS look simpler and for future operators we will have less replicated logic in the templates and will be rendering less clutter.

If this was a state/module for others to consume I'd agree present/absent is the way to go, but since this is for us and for making it easier to integrate this pattern in future operators for this mixed world I think I prefer to have simpler SLS and the big chunk of the logic in the python module.

I don't oppose to what you propose either.

Reply:
I agree it would mean more code on the SLSs, but (and this is a completely personal view) I think it would be better because

1) _explicit is better than implicit_, and it would be clearer what the code does.
2) I think it would follow more Salt's philosophy, where things are declared as _present_, _missing_ and so on instead of using _deus ex machina_ states that reconcile desired things. 

But, as I said, this is my personal take on this...

Reply:
i think personally that the saltstack way is to have `managed`, `absent` etc, logic.

I think yop, we we will have more rendundant code in SLS but if we gan simplicity is better to have it :rocket: